### PR TITLE
Kserve Dashboard Reconciler

### PR DESCRIPTION
**Pull request by ashley-o0o**
## Description
Add logic for configmap creation for Kserve metrics in ODH Model Controller
Closes: [RHOAIENG-3956](https://issues.redhat.com/browse/RHOAIENG-3956)

## How Has This Been Tested?
- install odh
- install the following DSC 
- ```spec:
  components:
    codeflare:
      managementState: Removed
    kserve:
      devFlags:
        manifests:
          - contextDir: config
            sourcePath: overlays/odh
            uri: 'https://github.com/opendatahub-io/kserve/tarball/master'
          - contextDir: config
            sourcePath: ''
            uri: 'https://github.com/ashley-o0o/odh-model-controller/tarball/kserveReconciler'
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: SelfSigned
        managementState: Managed
        name: knative-serving
    modelregistry:
      managementState: Removed
    trustyai:
      managementState: Removed
    ray:
      managementState: Removed
    kueue:
      managementState: Removed
    workbenches:
      managementState: Removed
    dashboard:
      managementState: Managed
    modelmeshserving:
      managementState: Removed
    datasciencepipelines:
      managementState: Removed
    trainingoperator:
      managementState: Removed
- create a data science project
- deploy a sample model for each supported ServingRuntime 
- ensure <modelname>-metrics-dashboard configmap exists, and contents are correct 
- ensure configmap is deleted when model is deleted. 
- ensure that the configmap cannot be changed by the user (the reconciler should revert changes) 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: config/base/params.env [hunk 1]
```diff
@@ -1,4 +1,4 @@
-odh-model-controller=quay.io/opendatahub/odh-model-controller:fast
+odh-model-controller=quay.io/rh-ee-amcent/odh-model-controller:latest
```
Comment (Jooho) [inline config/base/params.env:1]: is it the right url?

Comment (ashley-o0o) [reply config/base/params.env:1]: @VedantMahabaleshwarkar recommended I leave it as mine until others have tested it, then revert it back

File: controllers/reconcilers/kserve_metrics_dashboard_reconciler.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	servingRuntimeImage := runtime.Spec.Containers[0].Image
+	re := regexp.MustCompile(`/([^/@]+)[@:]`)
+	findImageName := re.FindStringSubmatch(servingRuntimeImage)
+	servingRuntime := findImageName[1]
+
+	var data []byte
+	switch servingRuntime {
+	case "openvino_model_server":
```
Comment (VedantMahabaleshwarkar) [inline controllers/reconcilers/kserve_metrics_dashboard_reconciler.go:111]: all the strings in the switch here should be defined as constants

File: controllers/reconcilers/kserve_metrics_dashboard_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	if err := ctrl.SetControllerReference(isvc, configMap, r.client.Scheme()); err != nil {
+		log.Error(err, "Unable to add OwnerReference to the Metrics Dashboard Configmap")
+		return nil, err
+	}
+
+	return configMap, nil
+}
+
+func substituteVariablesInQueries(data string, namespace string, name string) string {
+	return strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(data, "${namespace}", namespace), "${model_name}", name), "${rate_interval}", "1m")
```
Comment (VedantMahabaleshwarkar) [inline controllers/reconcilers/kserve_metrics_dashboard_reconciler.go:165]: define the rate interval value as a constant

File: controllers/reconcilers/kserve_metrics_dashboard_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+	case "vllm":
+		if vllmData == nil {
+			vllmData, err := os.ReadFile("vllm-metrics.json")
+			if err != nil {
+				log.Error(err, "Unable to load metrics dashboard template file:", err)
+				return nil, err
+			}
+			data = vllmData
+		}
```
Comment (VedantMahabaleshwarkar) [inline controllers/reconcilers/kserve_metrics_dashboard_reconciler.go:138]: we need a default case here for unsupported runtimes
```suggestion
		}
	default:
		log.V(1).Info("Metrics for runtime not supported.", "Runtime:", runtimeImageName)
		configMap := &corev1.ConfigMap{
			ObjectMeta: metav1.ObjectMeta{
				Name:      isvc.Name + ConfigMapSuffix,
				Namespace: isvc.Namespace,
			},
			Data: map[string]string{
				"supported": "false",
			},
		}
		if err := ctrl.SetControllerReference(isvc, configMap, r.client.Scheme()); err != nil {
			log.Error(err, "Unable to add OwnerReference to the Metrics Dashboard Configmap")
			return nil, err
		}
		return configMap, nil
```

File: controllers/reconcilers/kserve_metrics_dashboard_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	finaldata := substituteVariablesInQueries(string(data), isvc.Namespace, isvc.Name)
```
Comment (VedantMahabaleshwarkar) [inline controllers/reconcilers/kserve_metrics_dashboard_reconciler.go:141]: this method should also take the rate interval constant as an input parameter which is used to perform the substitution

File: controllers/utils/utils.go [hunk 6]
```diff
@@ -7,8 +7,9 @@ import (
 	"os"
 	"reflect"
 
-	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
 	"github.com/kuadrant/authorino/pkg/log"
+
+	kservev1beta1 "github.com/kserve/kserve/pkg/apis/serving/v1beta1"
```
Comment (VedantMahabaleshwarkar) [inline controllers/utils/utils.go:12]: does this need to be changed?

**Review by VedantMahabaleshwarkar — CHANGES_REQUESTED** [review]

File: Containerfile [hunk 7]
```diff
@@ -13,6 +13,9 @@ RUN go mod download
 COPY main.go main.go
 #COPY api/ api/
 COPY controllers/ controllers/
+COPY controllers/constants/ovms-metrics.json metrics_dashboards/ovms-metrics.json
+COPY controllers/constants/tgis-metrics.json metrics_dashboards/tgis-metrics.json
+COPY controllers/constants/vllm-metrics.json metrics_dashboards/vllm-metrics.json
```
Comment (spolti) [inline Containerfile:18]: where are these files?

Comment (ashley-o0o) [reply Containerfile:18]: They're from Killian's [PR 209](https://github.com/opendatahub-io/odh-model-controller/pull/209)

Comment (ashley-o0o) [reply Containerfile:18]: And just now added (forgot to put them in), but the caikit-metrics.json file is from [PR 218](https://github.com/opendatahub-io/odh-model-controller/pull/218)

Comment (spolti) [reply Containerfile:18]: okay, @KillianGolds's PR is merged, can you please rebase yours to use the files from there?

**Review by VedantMahabaleshwarkar — APPROVED** [review]
/lgtm

**Pull request merged**